src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/DiagnosticScopeCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,178 @@
+            }
+
+            var node = tree.FindNode(context.Span);
+            if (node is MethodDeclarationSyntax methodDeclarationSyntax)
+            {
+                var method = ModelExtensions.GetDeclaredSymbol(semanticModel, methodDeclarationSyntax) as IMethodSymbol;
+
+                if (method == null ||
+                    method.IsStatic ||
+                    method.IsAbstract ||
+                    method.ContainingType.TypeKind != TypeKind.Class)
+                {
+                    return;
+                }
+
+                var diagnosticScopeType = semanticModel.Compilation.GetTypeByMetadataName("Azure.Core.Pipeline.DiagnosticScope");
+                var clientDiagnosticsType = semanticModel.Compilation.GetTypeByMetadataName("Azure.Core.Pipeline.ClientDiagnostics");
+                var exceptionType = semanticModel.Compilation.GetTypeByMetadataName("System.Exception");
+
+                if (diagnosticScopeType == null || clientDiagnosticsType == null || exceptionType == null)
+                {
+                    return;
+                }
+
+                string clientDiagnosticsMember = null;
+                foreach (var member in method.ContainingType.GetMembers())
+                {
+                    if (
+                        (member is IFieldSymbol fieldSymbol && SymbolEqualityComparer.Default.Equals(clientDiagnosticsType, fieldSymbol.Type)) ||
+                        (member is IPropertySymbol propertySymbol && SymbolEqualityComparer.Default.Equals(clientDiagnosticsType, propertySymbol.Type))
+                    )
+                    {
+                        clientDiagnosticsMember = member.Name;
+                    }
+                }
+
+                if (clientDiagnosticsMember == null)
+                {
+                    return;
+                }
+
+                Task<Document> AddDiagnosticScope()
+                {
+                    var generator = SyntaxGenerator.GetGenerator(context.Document);
+
+                    var preconditions = new List<StatementSyntax>();
+                    var mainLogic = new List<SyntaxNode>();
+
+                    IEnumerable<StatementSyntax> statements;
+
+                    if (methodDeclarationSyntax.Body != null)
+                    {
+                        statements = methodDeclarationSyntax.Body.Statements;
+                    }
+                    else if (methodDeclarationSyntax.ExpressionBody != null)
+                    {
+                        statements = new [] { (StatementSyntax)generator.ReturnStatement(methodDeclarationSyntax.ExpressionBody.Expression) };
+                    }
+                    else
+                    {
+                        return Task.FromResult(context.Document);
+                    }
+
+                    foreach (var statement in statements)
+                    {
+                        if (mainLogic.Count == 0 &&
+                            IsPrecondition(statement))
```

Comment:
I would rename this to something closer to `ShouldStatementBeInsideScope` -- but better 🙂.

Reply:
Done